README.md
# Reveal Card

[![Pub Version](https://img.shields.io/pub/v/reveal_card)](https://pub.dartlang.org/packages/reveal_card)

Reveal Card widget helps you to create a card with a title, description, and a background image. Also you can add a button to the card. The card also contains a backside with a title, description and a background image. The card can be revealed up and the backside can be displayed.

## Sample

![](assets/20220504_150008_Reveal_Card.gif)

## Getting started

Sample Code

```
RevealCard(
          height: 350,
          width: 350,
          title: const Text(
            'HP Laptop',
            style: TextStyle(color: Colors.white, fontSize: 30),
          ),
          revealIcon: const Icon(
            Icons.more_vert,
            color: Colors.white,
          ),
          actionButton: FloatingActionButton(
            backgroundColor: Colors.redAccent,
            onPressed: () {},
            child: const Icon(Icons.add_shopping_cart),
          ),
          backgroundImage: const NetworkImage(
              'https://images.unsplash.com/photo-1525547719571-a2d4ac8945e2?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=764&q=80'),
          content: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text('Company', style: TextStyle(color: Colors.white, fontSize: 18)),
                Expanded(
                  child: Text(
                    'SAMSUNG',
                    style: TextStyle(color: Colors.white, fontSize: 18),
                    textAlign: TextAlign.right,
                  ),
                ),
              ],
            ),
            const SizedBox(
              height: 30,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text('Price', style: TextStyle(color: Colors.white, fontSize: 18)),
                Expanded(
                  child: Text(
                    '\$ 1349',
                    style: TextStyle(color: Colors.white, fontSize: 18),
                    textAlign: TextAlign.right,
                  ),
                ),
              ],
            ),
            const SizedBox(
              height: 30,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text('Rating', style: TextStyle(color: Colors.white, fontSize: 18)),
                Expanded(
                  child: Text(
                    '4.6',
                    style: TextStyle(color: Colors.white, fontSize: 18),
                    textAlign: TextAlign.right,
                  ),
                ),
              ],
            ),
          ],
          revealBackgroundImage: const NetworkImage(
              'https://images.unsplash.com/photo-1603302576837-37561b2e2302?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1468&q=80'),
          revealCloseIcon: const Icon(Icons.close, color: Colors.white),
          revealTitle: const Text(
            'Specifications',
            style: TextStyle(fontSize: 26, color: Colors.white),
          ),
          revealContent: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text(
                  'OS',
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
                Text(
                  'Windows 11',
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
              ],
            ),
            const SizedBox(
              height: 30,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text(
                  'Hard Drive',
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
                Text(
                  'SSD',
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
              ],
            ),
          ],
        ),
```

## Arguments


| Argument              | Type                 | Required / Optional                        |
| :---------------------- | ---------------------- | -------------------------------------------- |
| title                 | Text                 | Required                                   |
| revealIcon            | Icon                 | Optional                                   |
| content               | List Widget          | Required                                   |
| height                | double               | Optional (Default - 300)                   |
| width                 | double               | Optional (Default - 300)                   |
| actionButton          | FloatingActionButton | Optional                                   |
| backgroundImage       | ImageProvider        | Optional (Default - backgroundColor)       |
| opacity               | double (0 - 1)       | Optional (Default - 0.5)                   |
| backgroundColor       | Color                | Optional (Default - White)                 |
| revealTitle           | Text                 | Optional                                   |
| revealCloseIcon       | Icon                 | Optional                                   |
| revealContent         | List Widget          | Optional                                   |
| revealBackgroundColor | Color                | Optional (Default - White)                 |
| revealBackgroundImage | ImageProvider        | Optional (Default - revealBackgroundColor) |
| revealOpacity         | double (0 - 1)       | Optional (Default - 0.5)                   |
| revealDuration        | Duration             | Optional (Default - 200 milliseconds)      |

## Additional information

* Front side- Title, Content, Reveal Button, Action Button, Background Image.
* Back side- Title, Content, Close Button, Background Image.
* Action Button - Define function to be performed on user click.

On reveal button click, the back side will be revealed with a bottom up animation.

## Note

Enlarge the height and width of the card based on the content. Otherwise the contents will get overflowed.


CHANGELOG.md
## 1.0.0 - 2022-05-09

* Initial release of Reveal Card Widget

## 1.0.1 - 2022-07-04

* Increased SDK supported versions

test/reveal_card_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:reveal_card/reveal_card.dart';

void main() {
  test('adds one to input values', () {
    final revealCard = RevealCard();
    expect(revealCard.addOne(2), 3);
    expect(revealCard.addOne(-7), -6);
    expect(revealCard.addOne(0), 1);
  });
}


lib/src/reveal.dart
library reveal_card;

import 'package:flutter/material.dart';

class RevealCard extends StatefulWidget {
  const RevealCard({
    Key? key,
    required this.title,
    this.revealIcon,
    required this.content,
    this.backgroundImage,
    this.opacity,
    this.backgroundColor,
    this.revealTitle,
    this.revealCloseIcon,
    this.revealContent,
    this.revealBackgroundColor,
    this.revealBackgroundImage,
    this.revealOpacity,
    this.revealDuration,
    this.actionButton,
    this.width = 300,
    this.height = 300,
  }) : super(key: key);

  /// Title of the Front Card
  final Text? title;

  /// Icon to Reveal the Back Card
  final Icon? revealIcon;

  /// Content of the Front Card
  final List<Widget>? content;

  /// Background Image of the Front Card
  final ImageProvider? backgroundImage;

  /// Opacity of the Front Card
  final double? opacity;

  /// Background Color of the Front Card
  final Color? backgroundColor;

  /// Title of the Back Card
  final Text? revealTitle;

  /// Icon to close the Back Card
  final Icon? revealCloseIcon;

  /// Content of the Back Card
  final List<Widget>? revealContent;

  /// Background Color of the Back Card
  final Color? revealBackgroundColor;

  /// Background Image of the Back Card
  final ImageProvider? revealBackgroundImage;

  /// Opacity of the Back Card
  final double? revealOpacity;

  /// Duration of the Reveal Animation
  final Duration? revealDuration;

  /// Action Button of the Front Card
  final FloatingActionButton? actionButton;

  /// Width of the Card
  final double? width;

  /// Height of the Card
  final double? height;

  @override
  State<RevealCard> createState() => _RevealCardState();
}

class _RevealCardState extends State<RevealCard>
    with SingleTickerProviderStateMixin {
  /// Width of part to be revealed
  double revealWidth = 0;

  /// Height of part to be revealed
  double revealHeight = 0;

  @override
  Widget build(BuildContext context) {
    return Center(
      child: Material(
        elevation: 10.0,
        borderRadius: BorderRadius.circular(10.0),
        child: SizedBox(
          height: widget.height,
          width: widget.width,
          child: Stack(
            children: [
              widget.backgroundImage == null
                  ? Container(
                      decoration: BoxDecoration(
                        borderRadius: BorderRadius.circular(10.0),
                        color: widget.backgroundColor ?? Colors.white,
                      ),
                    )
                  : Container(
                      decoration: widget.backgroundImage == null
                          ? BoxDecoration(
                              borderRadius: BorderRadius.circular(10),
                              color: widget.backgroundColor ?? Colors.white)
                          : BoxDecoration(
                              borderRadius: BorderRadius.circular(10),
                              image: DecorationImage(
                                image: widget.backgroundImage!,
                                fit: BoxFit.fill,
                              ),
                            ),
                      child: Opacity(
                        opacity: widget.opacity ?? 0.5,
                        child: Container(
                          decoration: BoxDecoration(
                            borderRadius: BorderRadius.circular(10),
                            color: Colors.black,
                          ),
                        ),
                      ),
                    ),
              Positioned(
                top: 20,
                left: 20,
                right: 20,
                bottom: 0,
                child: Column(
                  children: [
                    Align(
                      alignment: Alignment.topLeft,
                      child: Row(
                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
                        children: [
                          Expanded(
                            child: widget.title ??
                                const Text(
                                  '',
                                  style: TextStyle(
                                    fontSize: 26,
                                    fontWeight: FontWeight.bold,
                                  ),
                                ),
                          ),
                          Row(
                            children: [
                              IconButton(
                                onPressed: () {
                                  setState(() {
                                    revealHeight =
                                        revealHeight == 0 ? widget.height! : 0;
                                  });
                                },
                                icon: widget.revealIcon ??
                                    const Icon(
                                      Icons.more_vert,
                                      color: Colors.black,
                                    ),
                              ),
                              SizedBox(
                                width: widget.actionButton == null ? 0 : 5,
                              ),
                              widget.actionButton ?? Container()
                            ],
                          ),
                        ],
                      ),
                    ),
                    const SizedBox(
                      height: 30,
                    ),
                    SizedBox(
                      height: widget.height! - 125,
                      child: SingleChildScrollView(
                        child: Column(
                          children: widget.content ?? [],
                        ),
                      ),
                    ),
                  ],
                ),
              ),
              Align(
                alignment: Alignment.bottomCenter,
                child: AnimatedContainer(
                  decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(10.0),
                    color: widget.revealBackgroundColor ?? Colors.white,
                  ),
                  height: revealHeight,
                  width: widget.width,
                  duration:
                      widget.revealDuration ?? Duration(milliseconds: 200),
                  child: Stack(
                    children: [
                      widget.revealBackgroundImage == null
                          ? Container()
                          : Container(
                              decoration: BoxDecoration(
                                color: Colors.white,
                                borderRadius: BorderRadius.circular(10),
                                image: widget.revealBackgroundImage == null
                                    ? null
                                    : DecorationImage(
                                        fit: BoxFit.fill,
                                        image: widget.revealBackgroundImage!,
                                      ),
                              ),
                              child: Opacity(
                                  opacity: widget.revealOpacity ?? 0.5,
                                  child: Container(
                                    decoration: BoxDecoration(
                                        borderRadius:
                                            BorderRadius.circular(10.0),
                                        color: Colors.black),
                                  )),
                            ),
                      Column(
                        children: [
                          Expanded(
                            child: SingleChildScrollView(
                              child: Column(
                                children: [
                                  Padding(
                                    padding: const EdgeInsets.only(
                                        left: 15.0, right: 15.0, top: 10.0),
                                    child: Row(
                                      mainAxisAlignment:
                                          MainAxisAlignment.spaceBetween,
                                      children: [
                                        Expanded(
                                          child: widget.revealTitle ??
                                              const Text(
                                                '',
                                                style: TextStyle(fontSize: 24),
                                              ),
                                        ),
                                        IconButton(
                                          padding: EdgeInsets.zero,
                                          onPressed: () {
                                            setState(() {
                                              revealHeight = 0;
                                            });
                                          },
                                          icon: widget.revealCloseIcon ??
                                              const Icon(
                                                Icons.close,
                                              ),
                                          iconSize: 26,
                                        ),
                                      ],
                                    ),
                                  ),
                                  const Divider(
                                    height: 10,
                                    color: Colors.grey,
                                  ),
                                  Padding(
                                    padding: const EdgeInsets.symmetric(
                                        horizontal: 15.0, vertical: 15.0),
                                    child: Column(
                                      children: widget.revealContent ?? [],
                                    ),
                                  ),
                                ],
                              ),
                            ),
                          )
                        ],
                      ),
                    ],
                  ),
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:reveal_card/reveal_card.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Reveal Card',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Reveal Card'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: RevealCard(
          height: 350,
          width: 350,
          title: const Text(
            'HP Laptop',
            style: TextStyle(color: Colors.white, fontSize: 30),
          ),
          revealIcon: const Icon(
            Icons.more_vert,
            color: Colors.white,
          ),
          actionButton: FloatingActionButton(
            backgroundColor: Colors.redAccent,
            onPressed: () {},
            child: const Icon(Icons.add_shopping_cart),
          ),
          backgroundImage: const NetworkImage(
              'https://images.unsplash.com/photo-1525547719571-a2d4ac8945e2?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=764&q=80'),
          content: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text('Company',
                    style: TextStyle(color: Colors.white, fontSize: 18)),
                Expanded(
                  child: Text(
                    'SAMSUNG',
                    style: TextStyle(color: Colors.white, fontSize: 18),
                    textAlign: TextAlign.right,
                  ),
                ),
              ],
            ),
            const SizedBox(
              height: 30,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text('Price',
                    style: TextStyle(color: Colors.white, fontSize: 18)),
                Expanded(
                  child: Text(
                    '\$ 1349',
                    style: TextStyle(color: Colors.white, fontSize: 18),
                    textAlign: TextAlign.right,
                  ),
                ),
              ],
            ),
            const SizedBox(
              height: 30,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text('Rating',
                    style: TextStyle(color: Colors.white, fontSize: 18)),
                Expanded(
                  child: Text(
                    '4.6',
                    style: TextStyle(color: Colors.white, fontSize: 18),
                    textAlign: TextAlign.right,
                  ),
                ),
              ],
            ),
          ],
          revealBackgroundImage: const NetworkImage(
              'https://images.unsplash.com/photo-1603302576837-37561b2e2302?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1468&q=80'),
          revealCloseIcon: const Icon(Icons.close, color: Colors.white),
          revealTitle: const Text(
            'Specifications',
            style: TextStyle(fontSize: 26, color: Colors.white),
          ),
          revealContent: [
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text(
                  'OS',
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
                Text(
                  'Windows 11',
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
              ],
            ),
            const SizedBox(
              height: 30,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: const [
                Text(
                  'Hard Drive',
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
                Text(
                  'SSD',
                  style: TextStyle(fontSize: 18, color: Colors.white),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
